### PR TITLE
BUG: Fix equality

### LINKED ISSUE #1599: BUG: Expression equality is broken and was never correct
In trying to track down the root cause of #1295, I found that the `all_equal` function here: https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L142-L143 is not correct and looking through the history it doesn't look like it ever was.

The current implementation is only comparing the first N arguments of `left` and `right`, where

```python
N = min(len(left), len(right))
```

```python
def all_equal(left, right, cache=None):
    if util.is_iterable(left):
        return util.is_iterable(right) and all(
            map(functools.partial(all_equal, cache=cache), left, right))

    if hasattr(left, 'equals'):
        return left.equals(right, cache=cache)
    return left == right
```

That is incorrect because it will falsely return `True` that a projection with a filter and any projection *without* a filter are equal:

```
In [9]: import ibis

In [10]: t = ibis.table([('string_col', 'string')], name='t')

In [11]: t[['string_col']][t.string_col == 'a'].equals(t[['string_col']])
Out[11]: True
```

This is fundamentally incorrect.

However, it appears that this bug is a dependency of fusing projections on a filtered `Selection`, meaning we're going to have to break backwards compatibility in the name of correctness for a release or two until actual fix is merged.

Here's an example of what ibis is currently spuriously allowing (for convenience, I'm assuming) that actually *depends* on equality behaving this way:

```
In [15]: t[t.string_col == 'a'][t[['string_col']]]
Out[15]:
ref_0
UnboundTable[table]
  name: t
  schema:
    string_col : string

Selection[table]
  table:
    Table: ref_0
  selections:
    Table: ref_0
  predicates:
    Equals[boolean*]
      left:
        string_col = Column[string*] 'string_col' from table
          ref_0
      right:
        Literal[string]
          a
```

I propose the following:

1. This is fixed for the release after 0.14
1. We bump the version of ibis to 1.0.0
1. We start following semantic versioning to provide our users with information about breaking changes
1. We reimplement the current behavior without the equality bug

### PR DESCRIPTION
**Pull request by cpcloud**
Closes #1599

This is a temporary fix that preserves most existing behavior until we can
refactor the compiler to separate compilation from optimization.

### CODE REVIEW

**Comment by cpcloud** [comment]
@kszucs can you review this PR when you get a chance?

File: ibis/expr/operations.py [hunk 1]
```diff
@@ -168,6 +175,22 @@ def aggregate(self, this, metrics, by=None, having=None):
...
+    def is_ancestor(self, other):
+        import ibis.expr.lineage as lin
+
+        if isinstance(other, ir.Expr):
+            other = other.op()
+
+        if self.equals(other):
+            return True
+
+        fn = lambda e: (lin.proceed, e.op())  # noqa: E731
```
Comment (kszucs) [inline ibis/expr/operations.py]: Perhaps It'd be useful to factor out this block to a `lin.flatten` function. I start to like even more this `lin.traverse` :)

File: ibis/expr/operations.py [hunk 2]
```diff
@@ -1862,22 +1885,27 @@ def root_tables(self):
...
-        expr = self.to_expr()
-        fn = lambda e: (lin.proceed, e.op())  # noqa: E731
-        for child in lin.traverse(fn, expr):
-            if child.equals(other):
-                return True
+        if not isinstance(other, type(self)):
+            return False
 
-        return False
+        return self.table.equals(other.table) and (
```
Comment (kszucs) [inline ibis/expr/operations.py]: The naming is a bit twisted, perhaps `empty_or_equal`?

File: ibis/expr/analysis.py [hunk 3]
```diff
@@ -774,7 +774,7 @@ def _check_fusion(self, root):
 
             # a * projection
             if (isinstance(val, ir.TableExpr) and
-                (self.parent.op().equals(val.op()) or
+                (self.parent.op().compatible_with(val.op()) or
```
Comment (kszucs) [inline ibis/expr/analysis.py]: Compatible_with is nice, but we seriously need to refactor the analysis code.

File: ibis/impala/tests/test_sql.py [hunk 4]
```diff
@@ -146,15 +148,14 @@ def test_nested_join_multiple_ctes():
   FROM t0
     INNER JOIN movies t5
       ON t0.`movieid` = t5.`movieid`
-),
-t2 AS (
+)
+SELECT t2.*
```
Comment (kszucs) [inline ibis/impala/tests/test_sql.py]: Reviewing SQL string is surprisingly hard, but it looks way better.

File: ibis/impala/tests/test_sql.py [hunk 5]
```diff
@@ -362,3 +363,154 @@ def test_multiple_filters2():
...
+  SELECT extract(`odate`, 'year') AS `year`, count(*) AS `count`
+  FROM t0
+  GROUP BY 1
+) t1
+  INNER JOIN (
+    SELECT extract(t0.`odate`, 'year') AS `year`, count(*) AS `count`
+    FROM t0
+    WHERE t0.`o_totalprice` > (
+      SELECT avg(t7.`o_totalprice`) AS `mean`
+      FROM t0 t7
```
Comment (kszucs) [inline ibis/impala/tests/test_sql.py]: At first it looks odd. If I'm correct `t0` is `tpch` and `t7` is `tpch.view()`, so it's actually correct.

**Comment by kszucs** [comment]
@cpcloud Should I merge?

File: ibis/expr/operations.py (hunk 2 above)
Comment (cpcloud) [reply ibis/expr/operations.py]: Sounds good, fixing ...

File: ibis/expr/operations.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/operations.py]: Cool, can you open an issue for that?

File: ibis/expr/analysis.py (hunk 3 above)
Comment (cpcloud) [reply ibis/expr/analysis.py]: yeah it's getting out of hand

File: ibis/expr/operations.py (hunk 1 above)
Comment (kszucs) [reply ibis/expr/operations.py]: Done https://github.com/ibis-project/ibis/issues/1612

File: ibis/expr/operations.py [hunk 6]
```diff
@@ -141,17 +141,27 @@ def has_resolved_name(self):
...
             itertools.starmap(
                 functools.partial(all_equal, cache=cache),
                 zip(left, right)
             )
         )
 
-    if hasattr(left, 'equals'):
-        return left.equals(right, cache=cache)
-    return left == right
+    return (hasattr(left, 'equals') and
```
Comment (kszucs) [inline ibis/expr/operations.py]: IMO the previous was more readable.

**Comment by kszucs** [comment]
The tests broke.

**Comment by cpcloud** [comment]
Yeah, using `and` was incorrect because `left == right` would be executed regardless. Fixed

**Comment by cpcloud** [comment]
Merging on green.

**Comment by kszucs** [comment]
@cpcloud I've recently updated the clickhouse-*-feedstocks to support python 3.7 and one of the builds was broken. Theoretically it's enough to restart the build, however I cannot do that - can You give me permissions?

**Comment by cpcloud** [comment]
@kszucs Is this why the `python35_test` job failed? I can restart once the other builds finish

**Comment by kszucs** [comment]
All of the builds should work now, let me know otherwise :)

**Comment by kszucs** [comment]
Re. I'm reproducing it locally, now that I have bandwith.

**Comment by kszucs** [comment]
@cpcloud Please restart your builds, it should(!!!) work now!

**Comment by cpcloud** [comment]
@kszucs Excellent, thanks for digging in.

**Pull request closed without merging**